### ISSUE TITLE
Make PropertiesAttribute recursive for submodules

### ISSUE THREAD

**Issue by cmeeren** [fscheck/FsCheck#394]
It would be great if `PropertiesAttribute` could work for all tests in all submodules of the module it is defined on. Currently it only works for tests directly inside the module it is defined on.

**Comment by ploeh** [comment]
I entirely understand and accept that people have different programming styles and all, and in any case, such a feature wouldn't affect me in any way, so I sort of don't care, but two observations...:

- As a general rule, I favour the Python philosophy that [Explicit is better than implicit](https://www.python.org/dev/peps/pep-0020). The use of both `Arb.register` and the `Arbitrary` property of attributes pull the FsCheck in an implicit direction that I find adds no value. I always recommend the use of combinators, `gen` expressions, and other type-safe alternatives.
- Hierarchies are rarely the best way to organise code...

**Comment by cmeeren** [comment]
> Hierarchies are rarely the best way to organise code...

A bit off-topic, but I'm eager to learn; how do you structure your unit tests in F#? I currently mirror my production hierarchies (each file has several modules, sometimes these modules have one more level of sub-modules), with one last layer of modules for each function I test, wherein I place all tests for that function:

```
Production:
  File.fs
    Module1
      Fun1
      Fun2
    Module2
      SubModule1
        Fun3
        ...
      SubModule2
        Fun4
        ...

Tests:
  File.fs
    Module1
      ModuleFun1
        Fun1Test1
        Fun1Test2
      ModuleFun2
        Fun2Test1
        Fun2Test2
    Module2
      SubModule1
        ModuleFun3
          Fun3Test1
        ...
      SubModule2
        ModuleFun4
          Fun4Test1
        ...
```

**Comment by ploeh** [comment]
I don't use much in the way of hierarchies in my production code, either. A hierarchy, by its very nature, is a way to organise things that excludes all other ways to organise the same things. I rarely find that there's only a single way to organise things, so I usually keep them unorganised, but clearly labelled.

I do put modules in namespaces, but again, I rarely turn them into hierarchies.

People often protest that they don't want to have hundreds of files with no organisation. I agree, but I think that the best way to avoid that problem is to not have hundreds of files at all. (And I don't cheat by putting thousands of lines of code into a single file; keeping code bases small is the most effective way to keep them manageable.)

**Comment by kurtschelfthout** [comment]
> As a general rule, I favour the Python philosophy that Explicit is better than implicit.

I agree with the philosophy, but Python??

**Comment by cmeeren** [comment]
I too first came to know that principle through PEP 20.

**Comment by kurtschelfthout** [comment]
Released in 2.16.3.
